### ISSUE TITLE
Add support for exporting requests to HTTP Archive format (.har)

### ISSUE THREAD

**Issue by JayNewstrom** [ChuckerTeam/chucker#401]
I'm not fixed on `.har`, but it's a relatively simple plain text format, that's well supported by many applications (including Charles Proxy).

We've run into many issues where people running our QA app are either not allowed, or not capable of running Charles Proxy and having the ability to view these requests there after would be a huge help to our debugging efforts.

My goal is to dump the entire `transactions` table into a single file that I can view in Charles Proxy.

I've done some initial digging, and I think the database contains all the information I would need in order to pull this off.

## :raising_hand: Do you want to develop this feature yourself?

- [x] Yes
- [ ] No

I'm willing to do the work for this, but it likely won't happen right away.

I'm also not convinced it's necessary to do this in Chucker core. I could fairly easily just look at the database in application code or a separate library that knows about Chuckers internal database and write the export code there. That being said, the reason I'm posting this here; I think this would be a feature that might be useful to other Churcker users as well.

**Comment by cortinico** [comment]
> I'm not fixed on .har, but it's a relatively simple plain text format, that's well supported by many applications (including Charles Proxy).

Just out of curiosity, have you considered other alternatives? 

> My goal is to dump the entire transactions table into a single file that I can view in Charles Proxy.

We currently already support the entire transaction list. I think the matter would be just to adapt to a desired format.

> I'm also not convinced it's necessary to do this in Chucker core

That depends on how complicated is to write this custom format.

**Comment by vbuberen** [comment]
I am curios why can't QA team use Chucker to view the data they need? 
You mention that Chucker already has all the data you need, so why would you not use Chucker? Would like to find out how we could improve the UI, especially since we plan to do some improvements in future.

The request itself sounds too specific, so it is up to community demand and simplicity of implementation to get it either into Chucker or not.

**Comment by JayNewstrom** [comment]
> > I'm not fixed on .har, but it's a relatively simple plain text format, that's well supported by many applications (including Charles Proxy).
> 
> Just out of curiosity, have you considered other alternatives?

I did look at the `.chls` format as well, but that seemed like it was going to be harder to implement, and Charles Proxy specific. `.har` is more widely adopted, and even what Chrome network tab uses as an import/export option.
`.har` is also supported by other tools such as Fiddler, Burp Proxy, etc.

> > My goal is to dump the entire transactions table into a single file that I can view in Charles Proxy.
> 
> We currently already support the entire transaction list. I think the matter would be just to adapt to a desired format.

I don't see an API to export/share the entire transactions list, what's the class/method called?
 
> I am curios why can't QA team use Chucker to view the data they need?
> You mention that Chucker already has all the data you need, so why would you not use Chucker? Would like to find out how we could improve the UI, especially since we plan to do some improvements in future.

They can use chucker, but the export format isn't easy to use. Some can't use Charles Proxy because of different organizational security personnel. 

My end goal is to have one button in the app that exports everything I need to debug an issue, including build information, network transactions, and console logs, as well as breadcrumbs on how the user got to where they are. This feature would end up being just one piece to that overall puzzle that I'd like to put together to make a bug report easy to file, easy to reproduce and debug.

**Comment by cortinico** [comment]
> I don't see an API to export/share the entire transactions list, what's the class/method called?

It was added in this PR https://github.com/ChuckerTeam/chucker/pull/305

**Comment by JayNewstrom** [comment]
> > I don't see an API to export/share the entire transactions list, what's the class/method called?
> 
> It was added in this PR #305

This is cool! And definitely useful! My goal would just be to take it one step further and put it into a format that can be imported/viewed by other tools.

It really looks like most of the work to do this is already done. It would mostly just be putting the existing room data model into the JSON format that `.har` expects.

**Comment by JayNewstrom** [comment]
I've started working on this, and made some really good progress. I've been able to import my exported har file in a few tools including Charles Proxy and https://toolbox.googleapps.com/apps/har_analyzer/

That being said, we're missing some data:
- We don't have the timings required (http://www.softwareishard.com/blog/har-12-spec/#timings) I'll need to look into how to get these out of OkHttp (likely https://square.github.io/okhttp/4.x/okhttp/okhttp3/-event-listener/ but that complicates things a good amount!)
- We don't have the request http version (http://www.softwareishard.com/blog/har-12-spec/#request) Again, I'll need to look into OkHttp on how to pull this out, but I don't see an obvious spot, if any with a quick search

If this is something that will be accepted as a contribution, I'll also need a little input on what the UI/API for this should be. I'd like to be able to get the contents of the file via API for my use case, but I think it would also be useful to have an export in the UI, similar to how we do for text and curl commands. I'm thinking something along these lines

![Screenshot_20200801-134015 copy](https://user-images.githubusercontent.com/713288/89108298-dbbc6e00-d3fc-11ea-9a3b-5b4994610ac9.png)

Here's my initial work in progress/proof of concept commit: https://github.com/JayNewstrom/chucker/commit/751910f0fa6f001f0a1c5eabc210daa5317acab0

I would clean this up (TODOs, UI hacks, etc) before PRing, if something along these lines would be accepted.

**Comment by cortinico** [comment]
> If this is something that will be accepted as a contribution

I think this is a great addition imho. On the other hand I'm a bit concerned by the two pieces that are missing:

- `timings` Those seems trickier to get. I was wondering if we can just omit them for now, and reiterate on this feature.
- `request http version` I think this is something that we should definitely add to the library anyway.

> If this is something that will be accepted as a contribution, I'll also need a little input on what the UI/API for this should be. I'd like to be able to get the contents of the file via API for my use case, but I think it would also be useful to have an export in the UI, similar to how we do for text and curl commands. I'm thinking something along these lines

Given that we have similar issues related to file saving/sharing, I would love to consolidate our approach here (see #405). Specifically we already have a 2 shares modes (cURL, Text) and 1 save file mode (binary mode). What's the reason why you ended up with sharing the file rather than saving it (asking out of curiosity).

**Comment by vbuberen** [comment]
I would omit `timings` for now. We have other intentions at the moment on redesigning entities we use and API. Also, it backs down to complexity vs community request. If we see more activity from the community requesting this - ok, let's do it. So far I see only one specific use case request, so not sure if it is worth the effort.

> * We don't have the request http version

You mean protocol version (http, http/2) ? If so we already have it and you can see it on a screenshot you attached.

**Comment by cortinico** [comment]
> You mean protocol version (http, http/2) ? If so we already have it and you can see it on a screenshot you attached.

True 👍 Thanks for pointing out. I've updated the comment

**Comment by JayNewstrom** [comment]
The `protocol` that's stored today is _only_ the response protocol. The har spec requires request and response protocol.

> What's the reason why you ended up with sharing the file rather than saving it (asking out of curiosity).

It was mostly just because that's what I could easily tie into for exporting all transactions. I'm not set on save/share, given direction, I'll gladly update either way.

We can give a simplified version of `timings` for phase 1, and iterate. What I have now satisfies the spec, but it's worth noting that it's wrong!

I'll complete the UI changes and finish what I can, with plans to iterate to make improvements. I'll also begin to start writing some tests.

As was pointed out in #405 we might also want to add binary responses to this. We could use base64 encoding (which har supports).

I think for a first pass though, I'm going to try and leave as much out as possible.

**Comment by JayNewstrom** [comment]
I've made some more progress here. https://github.com/ChuckerTeam/chucker/compare/develop...JayNewstrom:har

I think what I've done is enough for a first phase, but it's going to have some conflicts with one of my other open PRs #410  so I'd like to get that one in first, so I can resolve conflicts before submitting the PR.

I'd also like to do some more manual testing against my companies app (which is where I initially found the file provider issues). If anyone wants to poke at my branch, and provide UI feedback or anything like that I can update that as well.

**Comment by vbuberen** [comment]
Just merged #410 to unblock you. Will try to check your implementation for feedback soon. Have some issues with time availability now.

**Comment by gusdantas** [comment]
Dears,
I've updated the @JayNewstrom 's [solution](https://github.com/ChuckerTeam/chucker/pull/417) to add support for har file in [this PR](https://github.com/ChuckerTeam/chucker/pull/696).
Although, I don't know if I had to do something before, like open another issue. Do I need?
Thanks.

**Comment by vbuberen** [comment]
Hi there. Thanks for your contribution. We will take a look at your PR soon and provide feedback on it. No additional actions required so far.
